File: benches/ports/js/programs.js
// programs rows at n = 1 000 000, ported from acvus-interpreter-test/benches/programs.rs.
//
// Obligation across artifacts: `B`, `STEPS_PER_OUTER`, `TAPE`, the opcode
// numbers, `program`, `bracketTable` and the three machine bodies are that
// file's `rust_table`, `rust_scan` and `rust_call`, whose timed region includes
// building the program, decoding it and building the bracket table. The twin
// dispatches an eight-variant `enum` through `match`; the port dispatches the
// integer opcode through a dense `switch`, arms in the twin's order.
//
// Usage: node programs.js <row> [n] [reps]

const REPS = 3;
const N = 1000000;

const B = 32;
const STEPS_PER_OUTER = 6 * B + 6;
const MIN_OUTER = 1;
const TAPE = 30000;

const INC = 0;
const DEC = 1;
const LEFT = 2;
const RIGHT = 3;
const OPEN = 4;
const CLOSE = 5;
const OUT = 6;
const IN = 7;

function outerCount(n) {
  return Math.max(Math.floor(n / STEPS_PER_OUTER), MIN_OUTER);
}

function program(n) {
  const code = [];
  for (let k = 0; k < outerCount(n); k++) code.push(INC);
  code.push(OPEN);
  code.push(RIGHT);
  for (let k = 0; k < B; k++) code.push(INC);
  code.push(OPEN, DEC, RIGHT, INC, LEFT, CLOSE);
  code.push(LEFT, DEC, CLOSE);
  code.push(RIGHT, RIGHT, OUT);
  return code;
}

function decode(code) {
  return code.map((c) => (c <= OUT ? c : IN));
}

function bracketTable(code) {
  const jumps = new Array(code.length).fill(0);
  const stack = [];
  for (let i = 0; i < code.length; i++) {
    if (code[i] === OPEN) stack.push(i);
    if (code[i] === CLOSE && stack.length > 0) {
      const o = stack.pop();
      jumps[o] = i;
      jumps[i] = o;
    }
  }
  return jumps;
}

function bfTable(n) {
  const code = program(n);
  const prog = decode(code);
  const jumps = bracketTable(code);
  const tape = new Array(TAPE).fill(0);
  let pc = 0;
  let ptr = 0;
  let steps = 0;
  let out = 0;
  const plen = prog.length;
  while (pc < plen) {
    switch (prog[pc]) {
      case INC:
        tape[ptr] += 1;
        break;
      case DEC:
        tape[ptr] -= 1;
        break;
      case LEFT:
        ptr -= 1;
        break;
      case RIGHT:
        ptr += 1;
        break;
      case OPEN:
        if (tape[ptr] === 0) pc = jumps[pc];
        break;
      case CLOSE:
        if (tape[ptr] !== 0) pc = jumps[pc];
        break;
      case OUT:
        out += tape[ptr];
        break;
      default:
        break;
    }
    pc += 1;
    steps += 1;
  }
  return steps + out;
}

function bfScan(n) {
  const code = program(n);
  const prog = decode(code);
  const tape = new Array(TAPE).fill(0);
  let pc = 0;
  let ptr = 0;
  let steps = 0;
  let out = 0;
  const plen = prog.length;
  while (pc < plen) {
    switch (prog[pc]) {
      case INC:
        tape[ptr] += 1;
        break;
      case DEC:
        tape[ptr] -= 1;
        break;
      case LEFT:
        ptr -= 1;
        break;
      case RIGHT:
        ptr += 1;
        break;
      case OPEN:
        if (tape[ptr] === 0) {
          let d = 1;
          while (d > 0) {
            pc += 1;
            if (prog[pc] === OPEN) d += 1;
            else if (prog[pc] === CLOSE) d -= 1;
          }
        }
        break;
      case CLOSE:
        if (tape[ptr] !== 0) {
          let d = 1;
          while (d > 0) {
            pc -= 1;
            if (prog[pc] === OPEN) d -= 1;
            else if (prog[pc] === CLOSE) d += 1;
          }
        }
        break;
      case OUT:
        out += tape[ptr];
        break;
      default:
        break;
    }
    pc += 1;
    steps += 1;
  }
  return steps + out;
}

function bump(x, d) {
  return x + d;
}

function bfCall(n) {
  const code = program(n);
  const prog = decode(code);
  const jumps = bracketTable(code);
  const tape = new Array(TAPE).fill(0);
  let pc = 0;
  let ptr = 0;
  let steps = 0;
  let out = 0;
  const plen = prog.length;
  while (pc < plen) {
    switch (prog[pc]) {
      case INC:
        tape[ptr] = bump(tape[ptr], 1);
        break;
      case DEC:
        tape[ptr] = bump(tape[ptr], -1);
        break;
      case LEFT:
        ptr -= 1;
        break;
      case RIGHT:
        ptr += 1;
        break;
      case OPEN:
        if (tape[ptr] === 0) pc = jumps[pc];
        break;
      case CLOSE:
        if (tape[ptr] !== 0) pc = jumps[pc];
        break;
      case OUT:
        out += tape[ptr];
        break;
      default:
        break;
    }
    pc += 1;
    steps += 1;
  }
  return steps + out;
}

const ROWS = { "bf table": bfTable, "bf scan": bfScan, "bf call": bfCall };

const row = process.argv[2];
const n = process.argv[3] === undefined ? N : Number(process.argv[3]);
const reps = process.argv[4] === undefined ? REPS : Number(process.argv[4]);
const body = ROWS[row];
if (body === undefined) {
  console.error(`no row named ${JSON.stringify(row)}`);
  process.exit(1);
}

body(n);
const samples = [];
let value;
for (let r = 0; r < reps; r++) {
  const start = process.hrtime.bigint();
  value = body(n);
  samples.push(Number(process.hrtime.bigint() - start));
}
samples.sort((a, b) => a - b);
const median = samples[samples.length >> 1];
console.log(`${row}\t${(median / 1000).toFixed(1)}\t${value}`);

File: benches/ports/python/programs.py
"""programs rows at n = 1 000 000, ported from acvus-interpreter-test/benches/programs.rs.

Obligation across artifacts: `B`, `STEPS_PER_OUTER`, `TAPE`, the opcode
numbers, `program`, `bracket_table` and the three machine bodies are that
file's `rust_table`, `rust_scan` and `rust_call`, whose timed region includes
building the program, decoding it and building the bracket table. The twin
dispatches an eight-variant `enum` through `match`; Python has no jump table
for that, so the port dispatches an integer opcode through an `if`/`elif`
chain ordered as the twin's arms are.

Usage: programs.py <row> [n] [reps]
"""

import sys
import time

REPS = 3
N = 1_000_000

B = 32
STEPS_PER_OUTER = 6 * B + 6
MIN_OUTER = 1
TAPE = 30000

INC = 0
DEC = 1
LEFT = 2
RIGHT = 3
OPEN = 4
CLOSE = 5
OUT = 6
IN = 7


def outer_count(n):
    return max(n // STEPS_PER_OUTER, MIN_OUTER)


def program(n):
    code = [INC] * outer_count(n)
    code.append(OPEN)
    code.append(RIGHT)
    code.extend([INC] * B)
    code.extend([OPEN, DEC, RIGHT, INC, LEFT, CLOSE])
    code.extend([LEFT, DEC, CLOSE])
    code.extend([RIGHT, RIGHT, OUT])
    return code


def decode(code):
    return [c if c <= OUT else IN for c in code]


def bracket_table(code):
    jumps = [0] * len(code)
    stack = []
    for i, c in enumerate(code):
        if c == OPEN:
            stack.append(i)
        if c == CLOSE and stack:
            o = stack.pop()
            jumps[o] = i
            jumps[i] = o
    return jumps


def bf_table(n):
    code = program(n)
    prog = decode(code)
    jumps = bracket_table(code)
    tape = [0] * TAPE
    pc = 0
    ptr = 0
    steps = 0
    out = 0
    plen = len(prog)
    while pc < plen:
        op = prog[pc]
        if op == INC:
            tape[ptr] += 1
        elif op == DEC:
            tape[ptr] -= 1
        elif op == LEFT:
            ptr -= 1
        elif op == RIGHT:
            ptr += 1
        elif op == OPEN:
            if tape[ptr] == 0:
                pc = jumps[pc]
        elif op == CLOSE:
            if tape[ptr] != 0:
                pc = jumps[pc]
        elif op == OUT:
            out += tape[ptr]
        pc += 1
        steps += 1
    return steps + out


def bf_scan(n):
    code = program(n)
    prog = decode(code)
    tape = [0] * TAPE
    pc = 0
    ptr = 0
    steps = 0
    out = 0
    plen = len(prog)
    while pc < plen:
        op = prog[pc]
        if op == INC:
            tape[ptr] += 1
        elif op == DEC:
            tape[ptr] -= 1
        elif op == LEFT:
            ptr -= 1
        elif op == RIGHT:
            ptr += 1
        elif op == OPEN:
            if tape[ptr] == 0:
                d = 1
                while d > 0:
                    pc += 1
                    inner = prog[pc]
                    if inner == OPEN:
                        d += 1
                    elif inner == CLOSE:
                        d -= 1
        elif op == CLOSE:
            if tape[ptr] != 0:
                d = 1
                while d > 0:
                    pc -= 1
                    inner = prog[pc]
                    if inner == OPEN:
                        d -= 1
                    elif inner == CLOSE:
                        d += 1
        elif op == OUT:
            out += tape[ptr]
        pc += 1
        steps += 1
    return steps + out


def bump(x, d):
    return x + d


def bf_call(n):
    code = program(n)
    prog = decode(code)
    jumps = bracket_table(code)
    tape = [0] * TAPE
    pc = 0
    ptr = 0
    steps = 0
    out = 0
    plen = len(prog)
    while pc < plen:
        op = prog[pc]
        if op == INC:
            tape[ptr] = bump(tape[ptr], 1)
        elif op == DEC:
            tape[ptr] = bump(tape[ptr], -1)
        elif op == LEFT:
            ptr -= 1
        elif op == RIGHT:
            ptr += 1
        elif op == OPEN:
            if tape[ptr] == 0:
                pc = jumps[pc]
        elif op == CLOSE:
            if tape[ptr] != 0:
                pc = jumps[pc]
        elif op == OUT:
            out += tape[ptr]
        pc += 1
        steps += 1
    return steps + out


ROWS = {"bf table": bf_table, "bf scan": bf_scan, "bf call": bf_call}


def main():
    row = sys.argv[1]
    n = int(sys.argv[2]) if len(sys.argv) > 2 else N
    reps = int(sys.argv[3]) if len(sys.argv) > 3 else REPS
    body = ROWS[row]

    body(n)
    samples = []
    value = None
    for _ in range(reps):
        start = time.perf_counter_ns()
        value = body(n)
        samples.append(time.perf_counter_ns() - start)
    samples.sort()
    median = samples[len(samples) // 2]
    print(f"{row}\t{median / 1000.0:.1f}\t{value}")


main()
